packages/vite/src/node/cli.ts

```diff
@@ -263,31 +279,69 @@ cli
     `[boolean] force empty outDir when it's outside of root`,
   )
   .option('-w, --watch', `[boolean] rebuilds when modules have changed on disk`)
-  .action(async (root: string, options: BuildOptions & GlobalCLIOptions) => {
-    filterDuplicateOptions(options)
-    const { build } = await import('./build')
-    const buildOptions: BuildOptions = cleanOptions(options)
+  .option('--app', `[boolean] same as builder.entireApp`)
+  .action(
+    async (
+      root: string,
+      options: BuildEnvironmentOptions & BuilderCLIOptions & GlobalCLIOptions,
+    ) => {
+      filterDuplicateOptions(options)
+      const build = await import('./build')
 
-    try {
-      await build({
-        root,
-        base: options.base,
-        mode: options.mode,
-        configFile: options.config,
-        logLevel: options.logLevel,
-        clearScreen: options.clearScreen,
-        build: buildOptions,
-      })
-    } catch (e) {
-      createLogger(options.logLevel).error(
-        colors.red(`error during build:\n${e.stack}`),
-        { error: e },
+      const buildOptions: BuildEnvironmentOptions = cleanGlobalCLIOptions(
+        cleanBuilderCLIOptions(options),
       )
-      process.exit(1)
-    } finally {
-      stopProfiler((message) => createLogger(options.logLevel).info(message))
-    }
-  })
+
+      try {
+        const inlineConfig = {
+          root,
+          base: options.base,
+          mode: options.mode,
+          configFile: options.config,
+          logLevel: options.logLevel,
+          clearScreen: options.clearScreen,
+          build: buildOptions,
+          ...(options.app ? { builder: { entireApp: true } } : {}),
+        }
+        const patchConfig = (resolved: ResolvedConfig) => {
+          if (resolved.builder.entireApp) {
+            return
+          }
+          // Until the ecosystem updates to use `environment.config.build` instead of `config.build`,
+          // we need to make override `config.build` for the current environment.
+          // We can deprecate `config.build` in ResolvedConfig and push everyone to upgrade, and later
+          // remove the default values that shouldn't be used at all once the config is resolved
+          const environmentName = resolved.build.ssr ? 'ssr' : 'client'
+          ;(resolved.build as ResolvedBuildOptions) = {
+            ...resolved.environments[environmentName].build,
+          }
+        }
+        const config = await build.resolveConfigToBuild(
+          inlineConfig,
+          patchConfig,
+        )
+
+        if (config.builder.entireApp) {
+          const builder = await build.createBuilderWithResolvedConfig(
+            inlineConfig,
+            config,
+          )
+          await builder.buildApp()
+        } else {
+          // Single environment (client or ssr) build or library mode build
+          await build.buildWithResolvedConfig(config)
+        }
```

Comment:
Yes, the `vite build` CLI implementation was using `createBuilder()` and `build()` before. The current implementation was changed because we were resolving the config twice for a single environment build. We first did `createBuilder()` to resolve the config, and check if `builder.entireApp` was set, and then called `builder.buildApp()` or `build()`. We later changed this to resolve the config first, and then reuse that config. This isn't something that others are expected to do, it is just our CLI that needs to do one or the other depending on the config. Projects using the programatic API will know if they want to do a single environment or a full app build.

Reply:
Is there a reason needing to base on `resolvedConfig.builder.entireApp`, and then deciding whether to execute with the builder or single build later? Shouldn't `--app` means that it will always use the builder and do `builder.buildApp()`? Or do we want to support frameworks changing `build.entireApp` dynamically? (I feel like we shouldn't, like other sensitive properties, e.g. `root`, `build.ssr`, etc.

Reply:
The `root` can be changed by frameworks in the `config` hook, no? 🤔
About `build.entireApp`, ya, the idea was to give frameworks to be able to use Vite's CLI without having `--app` in the scripts. A clean `vite build` for the whole app, if this is the direction we end up going. We can force `--app` for now though (or remove it before the final release if we aren't sure about this).

Reply:
Changing `root` is kinda not great because our config is loaded using `root` before plugins could change them. And plugins could assume that the root is unchanged and use them early on.

I think it would be better for now to force `--app` if they want to use the builder. Otherwise it kinda misaligns the CLI and the JS API I mentioned above, where it implies `build()` could automatically call the builder if `build.entireApp` was dynamically set.

Reply:
The issue is how we end up with a clean `vite build` in the future. But ya, maybe it is fine to force the `--app` flag in the CLI for now though, we can always add this later on.